### PR TITLE
Changed to public all the getXXXResource methods of the cache adapters

### PR DESCRIPTION
**Pull request by pensiero**
This PR is a consequence of the issue #12

It's useful in case someone wants to obtain the resource and pass it to other modules that need the original resource, (avoid creating and configuring a new resource, but re-use the one previously configured).

### CODE REVIEW

**Comment by Maks3w** [comment]
Because this will be a public API we need specific tests for those methods.

**Comment by Maks3w** [comment]
Should we create a common method name `getResource` and for v3 add it to the interface?

**Comment by pensiero** [comment]
I thought about it, but there are some adapters, like the [WinCache](https://github.com/zendframework/zend-cache/blob/master/src/Storage/Adapter/WinCache.php) adapter, that doesn't have a `getWinCacheResource` method

**Comment by weierophinney** [comment]
> Should we create a common method name getResource and for v3 add it to the interface?

The changes as stated are to support #12, and allow getting at the underlying resource for an adapter only; they are completely BC. It's even arguable as to whether this would warrant a new minor version; since only visibility is changed, the only "new" feature is that you can access the methods via instances now.

I'm not entirely sure we would want to add a `getResource()` method to the adapter interface in the future. There may be cases where the adapter might be virtual (e.g., unit test mocking) or the underlying functionality does not use a resource (e.g., file system, black hole, memory and session adapters), which would make the method semantics ambiguous. You typically want the resource because you want to pass it to another object, and in such a case, you need to know exactly what you're retrieving; an `instanceof` check followed by calling the more specific method makes sense in those situations.

**Comment by Maks3w** [comment]
1. IMO it's a feature because new methods are available in public API. (What a feature mainly do)

2. I suggest to use getResource when possible as a name rule so those adapters share the same API methods.

**Comment by Maks3w** [comment]
@weierophinney This PR need test for ensure public API is respected in future releases

**Comment by weierophinney** [comment]
> I suggest to use getResource when possible as a name rule so those adapters share the same API methods.

@Maks3w — My point is that many of the existing adapters are not backed by PHP resources. And of those that are, some are true PHP resources, some are object instances. Additionally, documenting the return value in the interface would be tricky as a result; we could only specify `mixed`, which is not a great fit for an interface.

Since the method does not fit all adapters, and because the return value cannot be specified universally, it should *not* be in the interface.

> This PR need test for ensure public API is respected in future releases

Agreed. @pensiero — can you add tests for these, please? They can be as simple as just testing that the return value of calling the method on an instance returns something.

**Comment by pensiero** [comment]
@weierophinney ok, only that the return value is not null, or also that the return value is an instance of Redis/Memcached...?

**Comment by weierophinney** [comment]
@pensiero Check that it's an instance of the expected type (no need to also test `NotNull`, as an `InstanceOf` check validates that anyways).

**Comment by marc-mabe** [comment]
As already sayed in #12 the resources of all resource based storages **are** already public accessible from the resource manager. 
```php
$options  = $storage->getOptions();
$resource = $options->getResourceManager()->getResource($options->getResourceId());
```
The resource manager is where you configure and share resources.

Because of the resource manager part is a little bit hard to understand I would like to see this changed for ZF3 so the storage adapter will get injected with the real resource. But the resource manager contains lot of adapter specific code to instantiate resources from given configuration. This logic needs to be somewhere but in my opinion this could be better handled in the service manager.

What do you think?

**Comment by pensiero** [comment]
Do I have to do something to make this PR mergeable?

File: test/Storage/Adapter/RedisTest.php [hunk 1]
```diff
@@ -14,18 +14,6 @@
 
 class RedisTest extends CommonAdapterTest
 {
-    /**
-     *
-     * @var Cache\Storage\Adapter\RedisOptions
-     */
-    protected $_options;
```
Comment (Maks3w) [inline test/Storage/Adapter/RedisTest.php]: Don't remove this. Probably they are for to provide useful typehints for typechecker tools

**Comment by Maks3w** [comment]
Please rebase. Seems to have merge conflicts.

File: test/Storage/Adapter/RedisTest.php (hunk 1 above)
Comment (pensiero) [inline test/Storage/Adapter/RedisTest.php]: Done, did you mean for both `$_options` and `$_storage`, isn't it?

**Comment by pensiero** [comment]
Updated and resolved conflicts

**Comment by marc-mabe** [comment]
I don't know if it is a good idea to proxy this method. For which reason this is needed and why only proxy this one method and not methods like `RedisResourceManger::getMajorVersion()`. In the end you could proxy all methods but than you can also implement all directly within the adapter as it would no longer make sense but it will grow up the complexity a lot.

So what's really wrong with the following ?
```
$options  = $storage->getOptions();
$resource = $options->getResourceManager()->getResource($options->getResourceId());
```

**Comment by pensiero** [comment]
If it works, ok for me

**Comment by pensiero** [comment]
@weierophinney it's ok also for you not to merge this and use the solution that @Maks3w proposed?

**Comment by marc-mabe** [comment]
@weierophinney @Maks3w can I close this as `wontfix`

**Comment by weierophinney** [comment]
Absolutely.

**Pull request closed without merging**